From dbfc28704f8c7f857acb6eb49176dec54eec05ef Mon Sep 17 00:00:00 2001
From: beviu <contact@beviu.com>
Date: Mon, 5 Jan 2026 14:37:33 +0100
Subject: [PATCH 4/5] arch-x86: implement STUF and CLUF

Most of the code is taken from Berk's great work!
---
 src/arch/x86/isa/decoder/two_byte_opcodes.isa    |  2 ++
 .../interrupts_and_exceptions.py                 | 16 ++++++++++++++++
 2 files changed, 18 insertions(+)

diff --git a/src/arch/x86/isa/decoder/two_byte_opcodes.isa b/src/arch/x86/isa/decoder/two_byte_opcodes.isa
index e596aef08b..347fbb7efb 100644
--- a/src/arch/x86/isa/decoder/two_byte_opcodes.isa
+++ b/src/arch/x86/isa/decoder/two_byte_opcodes.isa
@@ -138,6 +138,8 @@
                         0x0: BasicOperate::SERIALIZE({{/*Nothing*/}},
                                                          IsSerializeAfter);
                         0x4: Inst::UIRET();
+                        0x3: Inst::STUF();
+                        0x2: Inst::CLUF();
                     }
                 }
                 0x6: Inst::LMSW(Ew);
diff --git a/src/arch/x86/isa/insts/general_purpose/control_transfer/interrupts_and_exceptions.py b/src/arch/x86/isa/insts/general_purpose/control_transfer/interrupts_and_exceptions.py
index bded4d28a0..151c6e9249 100644
--- a/src/arch/x86/isa/insts/general_purpose/control_transfer/interrupts_and_exceptions.py
+++ b/src/arch/x86/isa/insts/general_purpose/control_transfer/interrupts_and_exceptions.py
@@ -369,6 +369,22 @@ def macroop UIRET {
     # RSP := tempRsp;
     mov rsp, rsp, t8, dataSize=8
 };
+
+def macroop STUF {
+    .serialize_before
+    rdval t1, ctrlRegIdx("misc_reg::UserFaultEnabled")
+    limm t7, 1, dataSize=8
+    or t1, t1, t7, dataSize=8
+    wrval ctrlRegIdx("misc_reg::UserFaultEnabled"), t1
+};
+
+def macroop CLUF {
+    .serialize_before
+    rdval t1, ctrlRegIdx("misc_reg::UserFaultEnabled")
+    limm t7, ~(1), dataSize=8
+    and t1, t1, t7, dataSize=8
+    wrval ctrlRegIdx("misc_reg::UserFaultEnabled"), t1
+};
 """
 # let {{
 #    class INT(Inst):
-- 
2.52.0

